marklogic-data-hub/src/main/java/com/marklogic/hub/cli/Installer.java

```diff
@@ -0,0 +1,158 @@
+
+
+    public Installer(DataHub dataHub, HubConfigImpl hubConfig) {
+        this.dataHub = dataHub;
+        this.hubConfig = hubConfig;
+    }
+
+    public void install(File projectDir) {
+        logger.info("Installing DHF version " + hubConfig.getJarVersion());
+
+        initializeProject(projectDir);
+        dataHub.install();
+    }
+
+    /**
+     * The intended use case is that an executable DHF jar can be run from any directory, which means we need to first
+     * initialize a DHF project (specifically, generating the gradle.properties file) and then refresh HubConfig based
+     * on those properties and anything a client passed in via JVM props.
+     *
+     * @param projectDir
+     */
+    protected void initializeProject(File projectDir) {
+        logger.info("Initializing DHF into project directory: " + projectDir);
+
+        hubConfig.createProject(projectDir.getAbsolutePath());
+        hubConfig.initHubProject();
+
+        // Passing in "true" causes properties to be read from the current directory
+        // Include System properties so that a client can override e.g. mlHost/mlUsername/mlPassword via JVM props
+        hubConfig.refreshProject(System.getProperties(), true);
+
+        removeEmptyRangeIndexArrayFromFinalDatabaseFile(projectDir);
+        removePasswordsFromUserFiles(projectDir, hubConfig);
+    }
+
+    /**
+     * This is needed until DHFPROD-2547 is fixed. The empty range index array will remove any user-defined range
+     * indexes, which will also cause a reindex.
+     */
+    protected void removeEmptyRangeIndexArrayFromFinalDatabaseFile(File projectDir) {
+        final File file = new File(projectDir, "src/main/ml-config/databases/final-database.json");
+        if (file.exists()) {
+            ObjectNode node = readJsonFromFile(file);
+            if (node.has("range-element-index")) {
+                JsonNode json = node.get("range-element-index");
+                if (json instanceof ArrayNode) {
+                    ArrayNode array = (ArrayNode) json;
+                    if (array.size() == 0) {
+                        logger.info("Removing empty range-element-index array from " + file + " so any existing " +
+                            "user-defined range indexes are not lost, nor is an unnecessary reindex caused.");
+                        node.remove("range-element-index");
+                        writeJsonToFile(node, file);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * If a user already exists, we don't want to overwrite its password.
+     */
+    protected void removePasswordsFromUserFiles(File projectDir, HubConfigImpl hubConfig) {
+        ResourcesFragment existingUsers = new UserManager(hubConfig.getManageClient()).getAsXml();
+        File dir = new File(projectDir, "src/main/hub-internal-config");
+        if (dir.exists()) {
+            final Map<String, String> tokens = hubConfig.getAppConfig().getCustomTokens();
+
```

Comment:
```suggestion
if (tokens == null) {
   logger.warn(...)
   return;
}
```

Reply:
The token map wouldn't be null, as it's eagerly instantiated by AppConfig. AppConfig could be null, which would mean that the hubConfig.refresh call failed, and nothing's going to work, in which case an exception will naturally occur.